File: backend/app/schemas/progress.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class LessonProgressRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    lesson_id: uuid.UUID
    completed_at: datetime


class QuizScore(BaseModel):
    # Built by hand in the route, not from_attributes=True off QuizAttempt
    # directly — lesson_id comes from the attempt's quiz, not the attempt
    # row itself, so there's no single ORM object with both as attributes.
    quiz_id: uuid.UUID
    lesson_id: uuid.UUID
    correct_count: int
    graded_count: int


class CourseProgress(BaseModel):
    completed_lessons: int
    total_lessons: int
    # Beyond the roadmap's literal {completed_lessons, total_lessons,
    # quiz_scores} shape: per-lesson checkmarks on the course detail page
    # (M6.4) need to know *which* lessons are done, not just how many —
    # an aggregate count alone can't render that. Returned here instead of
    # a second per-lesson endpoint so the page stays a single request.
    completed_lesson_ids: list[uuid.UUID]
    quiz_scores: list[QuizScore]

File: frontend/src/api/progress.ts
import { apiRequest } from './client'

export interface LessonProgress {
  lesson_id: string
  completed_at: string
}

export interface QuizScore {
  quiz_id: string
  lesson_id: string
  correct_count: number
  graded_count: number
}

export interface CourseProgress {
  completed_lessons: number
  total_lessons: number
  completed_lesson_ids: string[]
  quiz_scores: QuizScore[]
}

export function completeLesson(lessonId: string): Promise<LessonProgress> {
  return apiRequest<LessonProgress>(`/lessons/${lessonId}/complete`, {
    method: 'POST',
  })
}

export function getCourseProgress(courseId: string): Promise<CourseProgress> {
  return apiRequest<CourseProgress>(`/courses/${courseId}/progress`)
}
